File: lib/trends/dispatch.ts
// Dispatch trend queries in parallel. Each result is streamed to the UI as soon as it lands.

import type { EventSink } from '../stream'
import type { TrendQuery, TrendResult, TrendQueryPlan, TrendDataBundle } from './types'
import { fetchGoogleTrends } from './google-trends'
import { fetchSteam } from './steam'
import { fetchTwitch } from './twitch'
import { fetchReddit } from './reddit'
import { fetchYoutube } from './youtube'
import { fetchAppStore } from './appstore'
import { fetchHuya } from './huya'
import { fetchDouyu } from './douyu'
import { fetchBilibili } from './bilibili'

async function runOne(q: TrendQuery): Promise<TrendResult> {
  switch (q.source) {
    case 'google-trends': return fetchGoogleTrends(q)
    case 'steam': return fetchSteam(q)
    case 'twitch': return fetchTwitch(q)
    case 'reddit': return fetchReddit(q)
    case 'youtube': return fetchYoutube(q)
    case 'appstore': return fetchAppStore(q)
    case 'huya': return fetchHuya(q)
    case 'douyu': return fetchDouyu(q)
    case 'bilibili': return fetchBilibili(q)
    default:
      // Unknown source: synthesize a not-ok TrendResult so the UI still renders something.
      return {
        ok: false, source: q.source, label: q.source, query: q,
        error: 'unknown source', fetchedAt: Date.now(), latencyMs: 0,
      }
  }
}

export async function dispatchTrendQueries(
  plan: TrendQueryPlan,
  sink: EventSink,
): Promise<TrendDataBundle> {
  const results: TrendResult[] = await Promise.all(
    plan.queries.map(async q => {
      const r = await runOne(q)
      sink.emit({ type: 'trend_result', result: r })
      return r
    }),
  )
  const bundle: TrendDataBundle = { plan, results }
  sink.emit({ type: 'trend_bundle', bundle })
  return bundle
}

File: server/stratsquad/trends/dispatch.py
"""Parallel dispatch of trend queries. Each result is streamed to the UI as it lands."""
from __future__ import annotations
import asyncio
from langgraph.config import get_stream_writer

from ..types import TrendQuery, TrendResult, TrendQueryPlan, TrendDataBundle
from .google_trends import fetch_google_trends
from .steam import fetch_steam
from .twitch import fetch_twitch
from .reddit import fetch_reddit
from .youtube import fetch_youtube
from .appstore import fetch_appstore
from .huya import fetch_huya
from .douyu import fetch_douyu
from .bilibili import fetch_bilibili


FETCHERS = {
    "google-trends": fetch_google_trends,
    "steam": fetch_steam,
    "twitch": fetch_twitch,
    "reddit": fetch_reddit,
    "youtube": fetch_youtube,
    "appstore": fetch_appstore,
    "huya": fetch_huya,
    "douyu": fetch_douyu,
    "bilibili": fetch_bilibili,
}


async def dispatch_trend_queries(plan: TrendQueryPlan) -> TrendDataBundle:
    writer = get_stream_writer()

    async def run_one(q: TrendQuery) -> TrendResult:
        fn = FETCHERS.get(q.source)
        if not fn:
            from time import time as _t
            return TrendResult(
                ok=False, source=q.source, label=q.source, query=q,
                error="unknown source", fetched_at=int(_t() * 1000), latency_ms=0,
            )
        r = await fn(q)
        writer({"type": "trend_result", "result": r.model_dump(by_alias=True, exclude_none=True)})
        return r

    results = await asyncio.gather(*[run_one(q) for q in plan.queries])
    bundle = TrendDataBundle(plan=plan, results=list(results))
    writer({"type": "trend_bundle", "bundle": bundle.model_dump(by_alias=True, exclude_none=True)})
    return bundle
